src/main/networking/mac-ca.ts

```diff
@@ -0,0 +1,175 @@
+  * undefined order.
+  *
+  * @param workdir A temporary directory where files can be written.
+  * @param keychain The full path to the keychain database to enumerate.
+  * @note This is only exported for testing.
+  */
+export async function* getFilteredCertificates(workdir: string, keychain: string, deps_: depInjection = {}): AsyncIterable<string> {
+  console.debug(`getting certificates from ${ keychain }...`);
+
+  const deps = _.defaults(deps_, { getPEMCertificates, spawnFile }) as Required<depInjection>;
+  const certIterator = deps.getPEMCertificates(workdir, keychain);
+
+  for await (const certPEM of certIterator) {
+    const cert = new crypto.X509Certificate(certPEM);
+    const certPath = path.join(workdir, 'cert.pem');
+
+    console.error('got cert', cert, cert.ca, cert.issuer, cert.subject, certPEM);
+    if (!cert.ca) {
+      console.debug('Skipping non-CA certificate', cert.subject);
+      continue;
+    }
+    if (cert.issuer !== cert.subject) {
+      console.debug('Skipping non-self-signed certificate', cert.subject);
+      continue;
+    }
+    await fs.promises.writeFile(certPath, certPEM, 'utf-8');
+    try {
+      await deps.spawnFile('/usr/bin/security', ['verify-cert', '-c', certPath, '-L', '-l', '-R', 'offline'], { stdio: console });
+    } catch (ex) {
+      console.debug('Skipping untrusted certificate', cert.subject);
+      continue;
+    }
+    yield certPEM;
+  }
+}
+
+/**
+ * Enumerate all system certificates as PEM, in undefined order.  This does not
+ * do the necessary processing to ensure they are valid for our use.
+ *
+ * @param workdir A temporary directory where files can be written.
+ * @note This is only exported for testing.
+ */
+export async function* getPEMCertificates(workdir: string, keychain?: string): AsyncIterable<string> {
+  // In order to avoid issues on machine with a very large number of certificates,
+  // write all certificates (in PEM format) to a file, and then read that file out.
+  const pemFilePath = path.join(workdir, 'all-certs.pem');
+  const pemFile = await fs.promises.open(pemFilePath, fs.constants.O_CREAT | fs.constants.O_TRUNC | fs.constants.O_WRONLY);
+  const pemFileStream = fs.createWriteStream(pemFilePath, { fd: pemFile });
+  const args = ['find-certificate', '-a', '-p'];
+
+  if (keychain) {
+    args.push(keychain);
+  }
+  await spawnFile('/usr/bin/security', args, { stdio: ['ignore', pemFileStream, console] });
+  await util.promisify((cb: (err?: Error | null | undefined) => void) => pemFileStream.close(cb))();
+  await pemFile.close();
+
+  let pemLines: string[] = [];
+
+  for await (const line of readFileByLine(pemFilePath)) {
+    pemLines.push(line);
+    if (line === '-----END CERTIFICATE-----') {
+      yield pemLines.join('\n');
+      pemLines = [];
+    }
+  }
+}
+
+/**
+ * Read the given file, returning one line at a time.
+ *
+ * @note This is only exported for testing.
```

Comment:
Can we consider if we really need to export this for testing? Seeing a note like this tells me that this is an internal implementation detail and that testing the output of functions that make use of this one should suffice.

Reply:
I just noticed this one here, but the same would apply for all other functions with the same note.